### DISCUSSION TITLE
Error on InfluxDB exporter: failed to convert OTLP metric to line protocol: unsupported non-monotonic sum

### DISCUSSION THREAD

**Discussion by ziomill** [open-telemetry/opentelemetry-collector-contrib#8341] [Q&A]
Hello,

We are trying to send metrics to InfluxDB through the **influxdb** exporter (provided by the contrib project) and we are experiencing this error:

```
2022-03-07T18:43:28.642Z	error	exporterhelper/queued_retry.go:186	Exporting failed. The error is not retryable. Dropping data.	{"kind": "exporter", "name": "influxdb", "error": "Permanent error: failed to convert OTLP metric to line protocol: unsupported non-monotonic sum 'runtime.jvm.memory.pool'", "dropped_items": 35}
2022-03-07T18:43:28.642684400Z go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
2022-03-07T18:43:28.642692400Z 	go.opentelemetry.io/collector@v0.46.0/exporter/exporterhelper/queued_retry.go:186
2022-03-07T18:43:28.642696700Z go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
2022-03-07T18:43:28.642700800Z 	go.opentelemetry.io/collector@v0.46.0/exporter/exporterhelper/metrics.go:134
2022-03-07T18:43:28.642704800Z go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
2022-03-07T18:43:28.642709100Z 	go.opentelemetry.io/collector@v0.46.0/exporter/exporterhelper/queued_retry_inmemory.go:121
2022-03-07T18:43:28.642713100Z go.opentelemetry.io/collector/exporter/exporterhelper/internal.consumerFunc.consume
2022-03-07T18:43:28.642717100Z 	go.opentelemetry.io/collector@v0.46.0/exporter/exporterhelper/internal/bounded_memory_queue.go:99
2022-03-07T18:43:28.642721100Z go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func2
2022-03-07T18:43:28.642725100Z 	go.opentelemetry.io/collector@v0.46.0/exporter/exporterhelper/internal/bounded_memory_queue.go:78
```

This is the OTel collector configuration:
```
receivers:
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:

exporters:
  # Send Metrics to InfluxDB
  influxdb:
    endpoint: http://influxdb:8086
    timeout: 500ms
    org: mycompany
    bucket: otel-metrics
    token: 78z0IzKhi7ik_cT9Dx-whP7Qe5Q9LO30d5b420cOR_rzKDJnI3I5gVnHfMlKWzjkPoSmOAbwycW82ENeli5BHA==
    metrics_schema: telegraf-prometheus-v2

    sending_queue:
      enabled: true
      num_consumers: 3
      queue_size: 10

    retry_on_failure:
      enabled: true
      initial_interval: 1s
      max_interval: 3s
      max_elapsed_time: 10s


### Configure the Collector
service:
  pipelines:
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [influxdb]
```

Do you have any suggestions?

Thank you

**Comment by jpkrohling** [comment]
cc @jacobmarble @8none1

**Reply by jacobmarble** [reply]
Hi @ziomill . Thanks for this, I'll take a look today.

**Comment by jacobmarble** [comment]
@ziomill I have reviewed the [OTel metrics data model spec](https://github.com/open-telemetry/opentelemetry-specification/blob/49a3a1add39c74c659467c8b2a44efc3e1bb23eb/specification/metrics/datamodel.md). It looks like we should treat cumulative, non-monotonic sums similar to a gauge value, for schema `telegraf-prometheus-v2`.

Please follow along here: https://github.com/influxdata/influxdb-observability/issues/50

**Reply by jacobmarble** [reply]
See also https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8348

**Comment by ziomill** [comment]
Hi @jacobmarble 

I have upgraded the **otel-collector** to the **opentelemetry-collector-contrib:0.47.0** (where the problem should have been fixed):

![image](https://user-images.githubusercontent.com/22731494/160103730-9f8909cf-c0bd-4224-8021-6e1bbef12cd7.png)
![image](https://user-images.githubusercontent.com/22731494/160104063-27b764c8-bfeb-4a0c-b11a-fe786f6250ae.png)

Unfortunately, I still have errors about metrics sending:

Using the **metrics_schema: telegraf-prometheus-v1**:

```bash
2022-03-25T10:19:59.925Z	error	exporterhelper/queued_retry.go:183	Exporting failed. The error is not retryable. Dropping data.	{"kind": "exporter", "name": "influxdb", "error": "Permanent error: failed to convert OTLP metric to line protocol: unsupported non-monotonic sum 'runtime.jvm.memory.pool'", "dropped_items": 40}
2022-03-25T10:19:59.926065000Z go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
2022-03-25T10:19:59.926072900Z 	go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/queued_retry.go:183
2022-03-25T10:19:59.926075900Z go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
2022-03-25T10:19:59.926080500Z 	go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/metrics.go:133
2022-03-25T10:19:59.926084400Z go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
2022-03-25T10:19:59.926086700Z 	go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/queued_retry_inmemory.go:118
2022-03-25T10:19:59.926088900Z go.opentelemetry.io/collector/exporter/exporterhelper/internal.consumerFunc.consume
2022-03-25T10:19:59.926091400Z 	go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/internal/bounded_memory_queue.go:99
2022-03-25T10:19:59.926093800Z go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func2
2022-03-25T10:19:59.926096200Z 	go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/internal/bounded_memory_queue.go:78
``` 

Using the **metrics_schema: telegraf-prometheus-v2**:
``` bash
otel-clct-container            | 2022-03-25T10:30:18.241Z       error   exporterhelper/queued_retry.go:183      Exporting failed. The error is not retryable. Dropping data.    {"kind": "exporter", "name": "influxdb", "error": "Perma
nent error: failed to convert OTLP metric to line protocol: unsupported non-monotonic sum 'runtime.jvm.memory.pool'", "dropped_items": 35}
otel-clct-container            | go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
otel-clct-container            |        go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/queued_retry.go:183
otel-clct-container            | go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
otel-clct-container            |        go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/metrics.go:133
otel-clct-container            | go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
otel-clct-container            |        go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/queued_retry_inmemory.go:118
otel-clct-container            | go.opentelemetry.io/collector/exporter/exporterhelper/internal.consumerFunc.consume
otel-clct-container            |        go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/internal/bounded_memory_queue.go:99
otel-clct-container            | go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func2
otel-clct-container            |        go.opentelemetry.io/collector@v0.47.0/exporter/exporterhelper/internal/bounded_memory_queue.go:78
```

Any suggestion?

Thank you

**Reply by jacobmarble** [reply]
That's correct. #8348 was merged just after the v0.47.0 release. In order to benefit from this fix, you'll need to either wait for the next release, or build opentelemetry-collector-contrib at https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/82a4889f3 or later.

**Reply by ziomill** [reply]
Thank you @jacobmarble 

Anyway, in the [v0.47.0 Release note](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases/tag/v0.47.0) the https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8348 is reported as merged:

![image](https://user-images.githubusercontent.com/22731494/160144252-c6480880-2420-4834-ad0b-7fb495faf0c7.png)

**Reply by pmm-sumo** [reply]
Yeah, apparently it's [part of the v0.47.0 release](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.47.0/exporter/influxdbexporter/go.mod#L7) looking at the component's go.mod

**Reply by jacobmarble** [reply]
My mistake x2.

#8462 was merged after v0.47.0, and is unrelated to this concern.

#8348 is, indeed, included in v0.47.0. The error `unsupported non-monotonic sum` persists because I tagged https://github.com/influxdata/influxdb-observability/releases/tag/otel2influx%2Fv0.2.15 one commit early.

```
commit 14db1a7807e294fd3fea47c084818cdb42f07710
Author: Jacob Marble <jacobmarble@influxdata.com>
Date:   Tue Mar 8 14:55:13 2022 -0800

    Merge pull request #51 from influxdata/jgm-non-monotonic-metric
    
    feat: add support for cumulative, non-monotonic sums

commit 04e1dbbfa96a856a9a916ef83df9fe7c3b04df2e (tag: otel2influx/v0.2.15, tag: influx2otel/v0.2.15, tag: common/v0.2.15)
Author: Jacob Marble <jacobmarble@influxdata.com>
Date:   Mon Mar 7 09:18:32 2022 -0800

    chore: update dependencies per dependabot suggestions
```

So, the fix for this issue is included in main, post-v0.47.0 due to #8462, but not in v0.47.0.

My apologies.

**Reply by pmm-sumo** [reply]
Thank you for checking this @jacobmarble 

v0.48.0 is just around the corner so we should have the fix released in a few days

**Reply by ziomill** [reply]
Just to tell you that with 0.48.0 it works as expected. Thank you!

P.S.: Now I would use Grafana with _InfluxDB_ data source. Do you know if there are prebuilt dashboards to import in order to visualize standard OpenTelemetry metrics?
